demos/induction_heads_multigpu.py

```diff
@@ -0,0 +1,285 @@
+    if rank == 0:
+        model = LlamaForCausalLM.from_pretrained(
+            model_path, torch_dtype=torch.bfloat16,
+        )
+    else:
+        with torch.device("meta"):
+            model = LlamaForCausalLM.from_pretrained(
+                model_path, torch_dtype=torch.bfloat16,
+            )
+    return model, config
+
+
+def fsdp_config(rank: int) -> dict[str: Any]:
+    """Return the config to be used by FSDP.
+
+    Args:
+    ----
+        rank: The worker rank
+
+    Returns:
+    -------
+        A dictionary containing keyword -> respective configuration.
+    """
+    def _module_init_fn(module: nn.Module) -> Callable:
+        """Return the function used for initializing modules on FSDP workers."""
+        return module.to_empty(
+            device=torch.cuda.current_device(),
+            recurse=False,
+        )
+
+    auto_wrap_policy = partial(
+        transformer_auto_wrap_policy,
+        transformer_layer_cls={
+            LlamaDecoderLayer,
+        },
+    )
+    sharding_strategy = ShardingStrategy.FULL_SHARD
+    device_id = torch.cuda.current_device()
+    sync_module_states = True
+    param_init_fn = _module_init_fn if rank != 0 else None
+    mp_policy = MixedPrecision(
+        param_dtype=torch.bfloat16,
+        buffer_dtype=torch.bfloat16,
+        reduce_dtype=torch.bfloat16,
+    )
+
+    config = {
+        "auto_wrap_policy": auto_wrap_policy,
+        "sharding_strategy": sharding_strategy,
+        "device_id": device_id,
+        "sync_module_states": sync_module_states,
+        "param_init_fn": param_init_fn,
+        "mixed_precision": mp_policy,
+    }
+    return config
+
+
+def calculate_induction_score(
+        config: LlamaConfig,
+        activation_dict: dict[str, torch.Tensor],
+        module_names: list[str],
+        sequence_length: int,
+    ) -> None:
+    """Calculate and save a heatmap of the induction scores for each attention
+    head.
+
+    Args:
+    ----
+        config: The model's HF config
+        activation_dict: Dictionary containing the activations retrieved using
+            FlexModel
+        module_names: A list of the module names to which we have attached
+            hooks
+        sequence_length: The sequence length of the prompt passed into the
+            model
+    """
+    induction_score_store = torch.zeros(
+        (
+            config.num_hidden_layers,
+            config.num_attention_heads,
+        ),
+        device=torch.cuda.current_device(),
+    )
+
+    for i, module_name in enumerate(module_names):
+        attn_maps = activation_dict[module_name][0].detach().to(
+            torch.cuda.current_device(),
+        )
+        induction_stripe = attn_maps.diagonal(
+            dim1=-2, dim2=-1, offset=1-sequence_length,
+        )
+        induction_score = einops.reduce(
+            induction_stripe, "batch head_index position -> head_index", "mean",
+        )
+        induction_score_store[i, :] = induction_score
```

Comment:
Add comments on these describing the operations that are being performed to calculate the induction scores (or put it in the docstring for this function)